engine/src/main/java/org/hibernate/validator/cfg/scriptengine/impl/OSGiScriptEvaluatorFactory.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.cfg.scriptengine.impl;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.script.ScriptEngineFactory;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import javax.validation.ValidationException;
+
+import org.hibernate.validator.cfg.scriptengine.ScriptEvaluator;
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.scriptengine.ScriptEvaluatorImpl;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+
+/**
+ * {@link org.hibernate.validator.cfg.scriptengine.ScriptEvaluatorFactory} suitable for OSGi environments. It is created
+ * based on {@code BundleContext} which is used to iterate through {@code Bundle}s and find all {@link ScriptEngineFactory}
+ * candidates.
+ *
+ * @author Marko Bekhta
+ */
+public class OSGiScriptEvaluatorFactory extends AbstractCacheableScriptEvaluatorFactory {
+
+	private final List<ScriptEngineManager> scriptEngineManagers;
+
+	public OSGiScriptEvaluatorFactory(BundleContext context) {
+		this.scriptEngineManagers = CollectionHelper.toImmutableList( findManagers( context ) );
+	}
+
+	@Override
+	protected ScriptEvaluator createNewScriptEvaluator(final String languageName) throws ScriptException {
+		return scriptEngineManagers.stream()
+				.map( manager -> manager.getEngineByName( languageName ) )
+				.filter( Objects::nonNull )
+				.map( engine -> new ScriptEvaluatorImpl( engine ) )
+				.findFirst()
+				.orElseThrow( () -> new ValidationException( String.format( "Wasn't able to find script evaluator for '%s'.", languageName ) ) );
+	}
+
+
+	private List<ScriptEngineManager> findManagers(BundleContext context) {
+		return findFactoryCandidates( context ).stream()
+				.map( className -> {
+					try {
+						return new ScriptEngineManager( Class.forName( className ).getClassLoader() );
```

Comment:
I don't think using `Class#forName()` is a good idea in OSGi. I'd expect people will rather plug-in just a very simple implementation of `ScriptEvaluatorFactory` which just returns an implementation they statically passed via the constructor.

Reply:
this factory is based on [1] and [2], and how script engines are handled there. But if as you said earlier - we will not support factory implementations for users to choose from - we probably wouldn't need this `OSGiScriptEvaluatorFactory` anyway :)

[1]https://svn.apache.org/repos/asf/felix/trunk/mishell/src/main/java/org/apache/felix/mishell/OSGiScriptEngineManager.java
[2]https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/osgi/impl/OSGiScriptEngineManager.java